types/decimal.go

```diff
@@ -0,0 +1,403 @@
+}
+
+// TODO panic if negative or if totalDigits < len(initStr)???
+// evaluate as an integer and return left padded string
+func (d Dec) ToLeftPaddedWithDecimals(totalDigits int8) string {
+	intStr := d.Int.String()
+	fcode := `%0` + strconv.Itoa(int(totalDigits)) + `s`
+	return fmt.Sprintf(fcode, intStr)
+}
+
+// TODO panic if negative or if totalDigits < len(initStr)???
+// evaluate as an integer and return left padded string
+func (d Dec) ToLeftPadded(totalDigits int8) string {
+	chopped := ChopPrecisionAndRound(d.Int)
+	intStr := chopped.String()
+	fcode := `%0` + strconv.Itoa(int(totalDigits)) + `s`
+	return fmt.Sprintf(fcode, intStr)
+}
+
+//     ____
+//  __|    |__   "chop 'em
+//       ` \     round!"
+// ___||  ~  _     -bankers
+// |         |      __
+// |       | |   __|__|__
+// |_____:  /   | $$$    |
+//              |________|
+
+// nolint - go-cyclo
+// Remove a Precision amount of rightmost digits and perform bankers rounding
+// on the remainder (gaussian rounding) on the digits which have been removed.
+func ChopPrecisionAndRound(d *big.Int) (chopped *big.Int) {
+
+	// remove the negative and add it back when returning
+	if d.Sign() == -1 {
+		// make d positive, compute chopped value, and then un-mutate d
+		d = d.Neg(d)
+		chopped = ChopPrecisionAndRound(d)
+		d = d.Neg(d)
+		chopped.Neg(chopped)
+		return chopped
+	}
+
+	// get the trucated quotient and remainder
+	quo, rem := big.NewInt(0), big.NewInt(0)
+	quo, rem = quo.QuoRem(d, precisionReuse, rem)
+
+	if rem.Sign() == 0 { // remainder is zero
+		return quo
+	}
+
+	//lenWhole := len(d.String())
+	//if quo.Sign() == 0 { // only the decimal places (ex. 0.1234)
+	//lenWhole++
+	//}
+	//lenQuo := len(quo.String())
+	//lenRem := len(rem.String())
+	//leadingZeros := lenWhole - (lenQuo + lenRem) // leading zeros removed from the remainder
+
+	//zerosToAdd := int64(lenRem - 1 + leadingZeros)
+	//multiplier := new(big.Int).Exp(big.NewInt(10), big.NewInt(zerosToAdd), nil)
+	//fiveLine := new(big.Int).Mul(big.NewInt(5), multiplier)
+
+	switch rem.Cmp(fivePrecision) {
+	case -1:
+		chopped = quo
+		return
+	case 1:
+		chopped = quo.Add(quo, big.NewInt(1))
+		return
+	default: // bankers rounding must take place
+		// always round to an even number
+		if quo.Bit(0) == 0 {
+			chopped = quo
+			return
+		}
+		chopped = quo.Add(quo, big.NewInt(1))
+		return
+	}
+}
+
+// RoundInt64 rounds the decimal using bankers rounding
+func (d Dec) RoundInt64() int64 {
+	chopped := ChopPrecisionAndRound(d.Int)
+	if !chopped.IsInt64() {
+		panic("Int64() out of bound")
+	}
+	return chopped.Int64()
+}
+
+// RoundInt round the decimal using bankers rounding
+func (d Dec) RoundInt() Int {
+	return NewIntFromBigInt(ChopPrecisionAndRound(d.Int))
+}
+
+//___________________________________________________________________________________
+
+// wraps d.MarshalText()
+func (d Dec) MarshalAmino() (string, error) {
+	if d.Int == nil {
+		d.Int = new(big.Int)
```

Comment:
Can we return a constant string here?

Reply:
confused - expand?

Reply:
That is what he was suggesting AFAICT.

Reply:
k - addressed - just define the zero values once now above